### PR TITLE
[DebugInfo] Report errors when DWARFUnitHeader::applyIndexEntry fails

### PR DESCRIPTION
**Pull request by bulbazord**
Motivation: LLDB is able to report errors about these scenarios whereas LLVM's DWARF parser only gives a boolean success/fail. I want to migrate LLDB to using LLVM's DWARFUnitHeader class, but I don't want to lose some of the error reporting, so I'm adding it to the LLVM class first.

### CODE REVIEW

**Comment by bulbazord** [comment]
I realize there are no tests for this currently, I'm working through that but may need some help. :)

**Comment by bulbazord** [comment]
Okay, I've added some tests and made sure that they actually test all the relevant code paths. This is good to review now! 😄

File: llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp [hunk 1]
```diff
@@ -334,21 +338,33 @@ Error DWARFUnitHeader::extract(DWARFContext &Context,
...
+                             Offset);
+
   auto *UnitContrib = IndexEntry->getContribution();
   if (!UnitContrib ||
       UnitContrib->getLength() != (getLength() + getUnitLengthFieldByteSize()))
-    return false;
+    return createStringError(errc::invalid_argument,
+                             "DWARF package unit at offset 0x%8.8" PRIx64
+                             " has an inconsistent index",
+                             Offset);
```
Comment (JDevlieghere) [inline llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:357]: Seems like this clause is covering two things that should be reported separately: the lack of a unit and the length of the contribution not matching the expected length. The latter could be part of the error message (e.g. expected x, actual y).

**Review by felipepiovezan — APPROVED** [review]
LGTM!
I find it surprising that, looking for the word "package" in the dwarf 5 spec yields the first result in section 7, which is usually reserved for the finer points of the data representation

**Review by dwblaikie — APPROVED** [review]
looks good to me

**Review by JDevlieghere — APPROVED** [review]
LGTM, thanks!

**Pull request merged**